### PR TITLE
More multiplication result converted to large type fixes

### PR DESCRIPTION
**Pull request by dalg24**
These are were the last ones in our headers.
After this we still need to look into benchmarks and tests.

It is a bit like warnings, they are quite annoying at times but if we don't strive to avoid them we might miss the one that matters.

### CODE REVIEW

File: core/src/Kokkos_MemoryPool.hpp [hunk 1]
```diff
@@ -196,9 +196,10 @@ class MemoryPool {
 
         stats.consumed_superblocks++;
         stats.consumed_blocks += block_used;
-        stats.consumed_bytes += block_used * block_size;
+        stats.consumed_bytes += static_cast<size_t>(block_used) * block_size;
         stats.reserved_blocks += block_count - block_used;
-        stats.reserved_bytes += (block_count - block_used) * block_size;
+        stats.reserved_bytes +=
+            static_cast<size_t>(block_count - block_used) * block_size;
```
Comment (dalg24) [inline core/src/Kokkos_MemoryPool.hpp:202]: Definitions here
https://github.com/kokkos/kokkos/blob/7d64c3d6581d2c7be03c48e8feea9408f89ca735/core/src/Kokkos_MemoryPool.hpp#L154-L156

File: core/src/impl/Kokkos_HostThreadTeam.hpp [hunk 2]
```diff
@@ -358,7 +358,7 @@ class HostThreadTeamData {
     int const num  = (m_work_end + m_work_chunk - 1) / m_work_chunk;
     int const part = (num + m_league_size - 1) / m_league_size;
 
-    m_work_range.first  = part * m_league_rank;
+    m_work_range.first  = static_cast<int64_t>(part) * m_league_rank;
```
Comment (dalg24) [inline core/src/impl/Kokkos_HostThreadTeam.hpp:361]: https://github.com/kokkos/kokkos/blob/7d64c3d6581d2c7be03c48e8feea9408f89ca735/core/src/impl/Kokkos_HostThreadTeam.hpp#L78-L80

File: core/src/Kokkos_ScratchSpace.hpp [hunk 3]
```diff
@@ -110,7 +110,7 @@ class ScratchMemorySpace {
     // Note: for team scratch m_offset is 0, since every
     // thread will get back the same shared pointer
     void* tmp           = m_iter + m_offset * size;
-    uintptr_t increment = size * m_multiplier;
+    uintptr_t increment = static_cast<uintptr_t>(size) * m_multiplier;
```
Comment (romintomasetti) [inline core/src/Kokkos_ScratchSpace.hpp:113]: Probably worth tested.

Comment (romintomasetti) [reply core/src/Kokkos_ScratchSpace.hpp:113]: So, is it a compiler warning/error you're fixing or rather some runtime issue?

Comment (dalg24) [reply core/src/Kokkos_ScratchSpace.hpp:113]: https://github.com/kokkos/kokkos/security/code-scanning/24

Comment (romintomasetti) [reply core/src/Kokkos_ScratchSpace.hpp:113]: I guess it's worth adding this `clang-tidy` check once you're done.

Comment (dalg24) [reply core/src/Kokkos_ScratchSpace.hpp:113]: Sure.  It looks like it was added in LLVM 13.

**Review by crtrott — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
Ignoring shared space failure in CUDA build

File: core/src/Kokkos_ScratchSpace.hpp (hunk 3 above)
Comment (romintomasetti) [reply core/src/Kokkos_ScratchSpace.hpp:113]: @dalg24 Followup then ?

Comment (dalg24) [reply core/src/Kokkos_ScratchSpace.hpp:113]: What are you asking?

Comment (dalg24) [reply core/src/Kokkos_ScratchSpace.hpp:113]: Sure we probably should work towards enabling all the bugprone-* checks, but no I am not working on it.

**Pull request merged**